### ISSUE TITLE
Expose IP address of peer somehow?

### ISSUE THREAD

**Issue by jkeljo** [sdwilsh/siobrultech-protocols#13]
I started working on some new features over at https://github.com/jkeljo/greeneye-monitor, and realized that I'm going to need the IP address/hostname of the monitor that I'm getting packets from. (I'm planning to make some requests to the GEM's API so I can detect which channels are net metering, what temperature unit is in effect, and so on.)

Any thoughts on how best to plumb that through?

**Comment by sdwilsh** [comment]
As I'm working on https://github.com/sdwilsh/brultech-serial2mqtt/, I'm running into the same issue!  Let me put some thought into this.

**Comment by sdwilsh** [comment]
It turns out that I'm given a `SerialTransport` via pyserial, so I don't quite have the same problem.

In your case, I think you just want to extend the existing protocol to ask the transport that is created for the connection who the peer is.  Something like this would work:

```python
from siobrultech_protocols.gem.protocol import PacketProtocol

class TcpPacketProtocol(PacketProtocol):
  def connection_made(self, transport: asyncio.BaseTransport):
    super().connection_made(transport)
    self._peername = transport.get_extra_info("peername")

  @property
  def peername(self) -> Optional[str]:
    return self._peername
```

You can then query `peername` off the protocol when you construct it in your `Monitor` class.  This probably won't work well though if you have multiple GEMs talking to the server.  In that case, we'd want to have a dictionary of connections that haven't been closed (plus also plugging into `connection_lost`), or maybe just known hosts?

We could provide a callback to connection made, but then you have to keep track of state yourself anyway, and the above solution might be easier to understand.  Thoughts?

**Comment by jkeljo** [comment]
Good point on `connection_lost`; one thing I didn't remember when looking at https://github.com/jkeljo/greeneye-monitor/pull/7 is that the `peername` logging that I previously did from `connection_made` and `connection_lost` was useful to Home Assistant; I should be providing notifications and doing that logging from the HA integration instead.

I think I'll try subclassing the protocol to detect connections coming and going, have `Monitors` be able to look up `Monitor` instances by both hostname and serial number, and instantiate new `Monitor` instances in `connection_made` rather than `_handle_packet` as is done today.

I know the GEM drops and re-opens its connection every few minutes; I'll have to think about what that looks like to clients of `Monitors`.

**Comment by sdwilsh** [comment]
You might be able to use keepalive settings to keep it open, but that depends on users setting that up right, which is hard.

All the IP shenanigans is why I'm actually opting to go with a direct serial connection and then dump it to mqtt :D

**Comment by jkeljo** [comment]
> All the IP shenanigans is why I'm actually opting to go with a direct serial connection and then dump it to mqtt :D

Ahhh, so that's what you're up to. :)

**Comment by jkeljo** [comment]
I haven't checked this yet, but I just realized that maybe I can send API commands to it over the connection that it's already established for sending packets. I wouldn't need an IP for that. Let me give that a shot...

**Comment by jkeljo** [comment]
Yep, that totally does work. Not sure it's actually simpler, though. :)

**Comment by jkeljo** [comment]
Between the API support that just landed (#15), and the ability to subclass the protocol, there are now two ways to deal with my original request. Closing.
